cacheops/redis_client.py

```diff
@@ -0,0 +1,215 @@
+                # these two help us out later for possible invalidation
+                'db_tables': [x for x, y in cond_dnfs],
+                'cond_dicts': [dict(i) for x, y in cond_dnfs for i in y]
+            }
+            return
+        except AttributeError:
+            # we are not in a transaction.
+            pass
+        self.load_script('cache_thing', LRU)(
+            keys=[cache_key],
+            args=[
+                pickle.dumps(data, -1),
+                json.dumps(cond_dnfs, default=str),
+                timeout
+            ]
+        )
+
+    def start_transaction(self):
+        self._local.cacheops_transaction_cache = ChainMap()
+
+    @handle_connection_failure
+    def commit_transaction(self):
+        for key, value in self._local.cacheops_transaction_cache.items():
+            self.load_script('cache_thing', LRU)(
+                keys=[key],
+                args=[
+                    pickle.dumps(value['data'], -1),
+                    json.dumps(value['cond_dnfs'], default=str),
+                    value['timeout']
+                ]
+            )
+
+    def end_transaction(self):
+        del self._local.cacheops_transaction_cache
+
+    def start_savepoint(self):
+        self._local.cacheops_transaction_cache.maps.append({})
+
+    def commit_savepoint(self):
+        self._local.cacheops_transaction_cache.maps[1].update(
+            self._local.cacheops_transaction_cache.maps[0]
+        )
+
+    def end_savepoint(self):
+        self._local.cacheops_transaction_cache.maps.pop(0)
+
+    @handle_connection_failure
+    def invalidate_dict(self, db_table, obj_dict):
+        try:
+            local_cache = self._local.cacheops_transaction_cache
+        except AttributeError:
+            # not in a transaction
+            pass
+        else:
+            # is this thing in our local cache?
+            for key, value in local_cache.items():
+                if 'db_tables' in value and 'cond_dicts' in value:
+                    for table, cond_dict in zip(value['db_tables'], value['cond_dicts']):
+                        # is this key for the table we are invalidating?
+                        if table == db_table:
+                            match = False
+                            for obj_key in set(obj_dict.keys()) & set(cond_dict.keys()):
+                                if obj_dict[obj_key] != cond_dict[obj_key]:
+                                    break
+                            else:
+                                match = True
+                            if match or not cond_dict:
+                                # deep delete, to deal with savepoints in the cache
+                                for mapping in local_cache.maps:
+                                    if key in mapping:
+                                        del mapping[key]
+
+        self.load_script('invalidate')(args=[
+            db_table,
+            json.dumps(obj_dict, default=str)
+        ])
```

Comment:
Should not be called right away if in transaction, but queued:
```python
self._local.queue.append((self.load_script('invalidate'), dict(args=[...])))
```
And then run the queue on commit.